packages/vitest/src/integrations/chai/poll.ts

```diff
@@ -0,0 +1,75 @@
+import * as chai from 'chai'
+import type { ExpectStatic } from '@vitest/expect'
+import { getSafeTimers } from '@vitest/utils'
+
+// these matchers are not supported because they don't make sense with poll
+const unsupported = [
+  // .poll is meant to retry matchers until they succeed, and
+  // snapshots will always succeed as long as the poll method doesn't thow an error
+  // in this case using the `vi.waitFor` method is more appropriate
+  'matchSnapshot',
+  'toMatchSnapshot',
+  'toMatchInlineSnapshot',
+  'toThrowErrorMatchingSnapshot',
+  'toThrowErrorMatchingInlineSnapshot',
+  // toThrow will never succeed because we call the poll callback until it doesn't throw
+  'throws',
+  'Throw',
+  'throw',
+  'toThrow',
+  'toThrowError',
+  // these are not supported because you can call them without `.poll`,
+  // we throw an error inside the rejects/resolves methods to prevent this
+  // rejects,
+  // resolves
+]
+
+export function createExpectPoll(expect: ExpectStatic): ExpectStatic['poll'] {
+  return function poll(fn, options = {}) {
+    const { interval = 50, timeout = 1000, message } = options
+    const STACK_TRACE_ERROR = new Error('STACK_TRACE_ERROR')
+    // @ts-expect-error private poll access
+    const assertion = expect(null, message).withContext({ poll: true }) as Assertion
+    const proxy: any = new Proxy(assertion, {
+      get(target, key, receiver) {
+        const result = Reflect.get(target, key, receiver)
+
+        if (typeof result !== 'function')
+          return result instanceof chai.Assertion ? proxy : result
+
+        if (typeof key === 'string' && unsupported.includes(key))
+          throw new SyntaxError(`expect.poll() is not supported in combination with .${key}(). Use vi.waitFor() if your assertion condition is unstable.`)
+
+        return function (this: any, ...args: any[]) {
+          return new Promise((resolve, reject) => {
+            let intervalId: any
+            let lastError: any
+            const { setTimeout } = getSafeTimers()
+            setTimeout(() => {
+              clearTimeout(intervalId)
+              reject(copyStackTrace(new Error(`Matcher did not succeed in ${timeout}ms`, { cause: lastError }), STACK_TRACE_ERROR))
+            }, timeout)
+            const check = async () => {
+              try {
+                chai.util.flag(this, 'object', await fn())
```

Comment:
If the `fn` throws, the printed error message points to the `expect` call. Should we instead capture the error thrown by `fn` and show its source in the code frame?

<details>

```ts
import { expect, test } from "vitest";
import { sum } from "../src/example";

test("testing", async () => {
  await expect.poll(() => sum(1, 2)).toBe(3);
});
```

```ts
export function sum(a, b) {
  a.b.nonExistingField;
  return a + b + 1;
}
```

```
 FAIL  test/third.test.ts > third
Error: Matcher did not succeed in 1000ms
 ❯ test/third.test.ts:5:16
      3| 
      4| test("third", async () => {
      5|   await expect.poll(() => sum(1, 2)).toBe(3);
       |                ^
      6| });
      7| 

Caused by: TypeError: Cannot read properties of undefined (reading 'nonExistingField')
 ❯ Module.sum src/example.ts:6:7
 ❯ test/third.test.ts:5:27
```

</details>

Showing `expect` as erroneous lines is correct below when `result.call()` fails.

Reply:
Not sure how this would work. Do you want to replace the first error callstack? Do you want to rethrow the original error? How do you see the end result?

Reply:
Something like this:

```ts
return new Promise((resolve, reject) => {
  let intervalId: any
  let fnError: any
  let lastError: any
  const { setTimeout } = getSafeTimers()
  setTimeout(() => {
    clearTimeout(intervalId)
    const error = fnError || copyStackTrace(new Error(`Matcher did not succeed in ${timeout}ms`, { cause: lastError }), STACK_TRACE_ERROR)
    reject(error)
  }, timeout)
  const check = async () => {
    let output

    try {
      output = await fn()
      fnError = null
    }
    catch (err) {
      fnError = err
      intervalId = setTimeout(check, interval)
      return
    }

    try {
      chai.util.flag(this, 'object', output)
      resolve(await result.call(this, ...args))
    }
    catch (err) {
      lastError = err
      intervalId = setTimeout(check, interval)
    }
  }
  check()
})
```

Now it will point to the errorneous line if `fn()` threw error:

```ts
TypeError: Cannot read properties of undefined (reading 'nonExistingField')
 ❯ Module.sum src/example.ts:6:7
      4| 
      5| export function sum(a, b) {
      6|   a.b.nonExistingField;
       |       ^
      7|   return a + b + 1;
      8| }
 ❯ test/third.test.ts:5:27
 ```
 
 And when the `fn()` successfully returns a value but `expect().<matcher>()` fails:
 
 ```ts
 Error: Matcher did not succeed in 1000ms
 ❯ test/third.test.ts:5:16
      3| 
      4| test("third", async () => {
      5|   await expect.poll(() => sum(1, 2)).toBe(3);
       |                ^
      6| });
      7| 
```

Reply:
But this is not entirely correct - the error happened because it didn't return a valid value before the interval timed out, so the `Matcher did not succeed` is the correct error here - it doesn't matter that it happened inside the callback

Reply:
So it's expected that in `await expect.poll(() => sum())`, if there's an error inside `sum()`, it's expected to swallow the error and print just `'Matcher did not succeed in 1000ms'`. Not show that `sum()` actually threw an error? It's not the `.toBe()` that failed here.

Reply:
What do you mean by "swallow"? It is printed in `cause`

Reply:
I do agree that it's not that visible, but we also need to show the timeout error somewhere

Reply:
Interestingly enough, the playwright doesn't even allow errors in `poll`:

```js
  1) playwright.test.js:2:5 › basic test 

    Error: not yet

      3 |   await expect.poll(() => {
      4 |     throw new Error('not yet')
    > 5 |   }).toBe(3)
        |      ^
      6 | })
```

Reply:
Right, sure it's visible in stack trace and even the error message of original error is printed there. I guess it's just the code frame that's bothering me.

I'm fine with the current implementation as is. 👍

Reply:
On the other hand, it would be more consistent to go with the `vi.waitFor` implementation where we always reject the last error